sys/net/network_layer/ng_ipv6/netif/ng_ipv6_netif.c

```diff
@@ -440,14 +519,181 @@ static ipv6_addr_t *_match_prefix(kernel_pid_t pid, const ipv6_addr_t *addr,
     return NULL;
 }
 
-ipv6_addr_t *ng_ipv6_netif_match_prefix(kernel_pid_t pid, const ipv6_addr_t *prefix)
+/**
+ * @brief Determines the scope of the given address.
+ *
+ * @param[in] addr              The IPv6 address to check.
+ * @param[in] maybe_multicast   False if @p addr is definitely no multicast
+ *                              address, true otherwise.
+ *
+ * @return The scope of the address.
+ *
+ * @pre address is not loopback or unspecified.
+ * see http://tools.ietf.org/html/rfc6724#section-4
+ */
+static uint8_t _get_scope(const ipv6_addr_t *addr, const bool maybe_multicast)
+{
+    if (maybe_multicast && ipv6_addr_is_multicast(addr)) {
+        return (addr->u8[1] & 0x0f);
+    }
+    else if (ipv6_addr_is_link_local(addr)) {
+        return IPV6_ADDR_MCAST_SCP_LINK_LOCAL;
+    }
+    else if (ipv6_addr_is_site_local(addr)) {
+        return IPV6_ADDR_MCAST_SCP_SITE_LOCAL;
+    }
+    else {
+        return IPV6_ADDR_MCAST_SCP_GLOBAL;
+    }
+}
+
+/** @brief Find the best candidate among the configured addresses
+ *          for a certain destination address according to the 8 rules
+ *          specified in RFC 6734, section 5.
+ * @see http://tools.ietf.org/html/rfc6724#section-5
+ *
+ * @param[in] iface              The interface for sending.
+ * @param[in] dest               The destination IPv6 address.
+ * @param[in, out] candidate_set The preselected set of candidate addresses as
+ *                               a bitfield.
+ *
+ * @pre @p dest is not unspecified.
+ *
+ * @return The best matching candidate found on @p iface, may be NULL if none
+ *         is found.
+ */
+static ipv6_addr_t *_source_address_selection(ng_ipv6_netif_t *iface,
+                                      const ipv6_addr_t *dest,
+                                      uint8_t *candidate_set)
 {
-    return _match_prefix(pid, prefix, false);
+    /* create temporary set for marking candidates in the corresponding rules.
+     * Instead of creating several sets, one big set is allocated, in order to
+     * improve code size.
+     * According to rule 2:
+     * First `n` bits mark addresses that match the scope of `dest`.
+     * Bits `n` to `2n`mark addresses that have a smaller scope.
+     * According to rule 3:
+     * Bits `2n` to `3n` mark address that are preferred.
+     *
+     * Bits `3n` to `5n` mark two set of addresses that match several winning
+     * rules and need a tiebreaker.
+     */
+    uint8_t winner_set[NG_IPV6_NETIF_ADDR_NUMOF];
```

Comment:
Relates to my question above. If the candidate set can contain addresses from different interfaces, than using `NG_IPV6_NETIF_ADDR_NUMOF` may not be enough.